src/OccupancyGrid.cc

```diff
@@ -0,0 +1,362 @@
+  double originX;
+  double originY;
+
+  std::vector<CellState> gridData;
+
+  // Helper to get the 1D index from 2D grid coordinates
+  int GetIndex(int _gridX, int _gridY) const {
+    return _gridY * this->widthCells + _gridX;
+  }
+
+  // Constructor for Impl
+  Impl(double _resolutionMeters, int _widthCells, int _heightCells,
+    double _originX, double _originY)
+    : resolutionMeters(_resolutionMeters),
+      widthCells(_widthCells),
+      heightCells(_heightCells),
+      originX(_originX),
+      originY(_originY),
+      gridData(_widthCells * _heightCells, CellState::Unknown)
+  {}
+
+  // IsValidGridCoordinate is also needed internally for Impl operations
+  bool IsValidGridCoordinateImpl(int gridX, int gridY) const {
+    return gridX >= 0 && gridX < this->widthCells &&
+      gridY >= 0 && gridY < this->heightCells;
+  }
+
+  // Get and Set CellState are also needed internally
+  CellState GetCellStateImpl(int gridX, int gridY) const
+  {
+    if (this->IsValidGridCoordinateImpl(gridX, gridY))
+    {
+      return this->gridData[GetIndex(gridX, gridY)];
+    }
+    return CellState::Unknown;
+  }
+
+  void SetCellStateImpl(int gridX, int gridY, CellState state)
+  {
+    if (this->IsValidGridCoordinateImpl(gridX, gridY))
+    {
+      this->gridData[GetIndex(gridX, gridY)] = state;
+    }
+  }
+};
+
+/////////////////////////////////////////////////
+OccupancyGrid::OccupancyGrid(double resolutionMeters, int widthCells,
+  int heightCells, double originX, double originY)
+    : pImpl(std::make_unique<Impl>(
+      resolutionMeters, widthCells, heightCells, originX, originY))
+{
+}
+
+/////////////////////////////////////////////////
+OccupancyGrid::~OccupancyGrid() = default;
+
+/////////////////////////////////////////////////
+OccupancyGrid::OccupancyGrid(OccupancyGrid&& other) noexcept
+    : pImpl(std::move(other.pImpl))
+{}
+
+/////////////////////////////////////////////////
+OccupancyGrid& OccupancyGrid::operator=(OccupancyGrid&& other) noexcept
+{
+  if (this != &other)
+  {
+    pImpl = std::move(other.pImpl);
+  }
+  return *this;
+}
+
+/////////////////////////////////////////////////
+bool OccupancyGrid::WorldToGrid(double worldX, double worldY,
+  int& gridX, int& gridY) const
+{
+  gridX = static_cast<int>(
+    std::round((worldX - this->pImpl->originX) / this->pImpl->resolutionMeters)
+  );
+  gridY = static_cast<int>(
+    std::round((worldY - this->pImpl->originY) / this->pImpl->resolutionMeters)
+  );
+  return this->pImpl->IsValidGridCoordinateImpl(gridX, gridY);
+}
+
+/////////////////////////////////////////////////
+void OccupancyGrid::GridToWorld(int gridX, int gridY,
+  double& worldX, double& worldY) const
+{
+  worldX = gridX * this->pImpl->resolutionMeters + this->pImpl->originX;
+  worldY = gridY * this->pImpl->resolutionMeters + this->pImpl->originY;
+}
+
+/////////////////////////////////////////////////
+bool OccupancyGrid::IsValidGridCoordinate(int gridX, int gridY) const
```

Comment:
Not sure why the extra function is needed. Could it not be implemented here? Same with `GetCellState` and `SetCellState`

Reply:
d82ae46